sw/device/tests/sram_ctrl_main_scrambled_access_test.c

```diff
@@ -0,0 +1,179 @@
+
+/**
+ * MAIN SRAM scrambling test buffer.
+ *
+ * This buffer is initialized with a known pattern, which will be overwritten
+ * with a pseudo-random data as the result of MAIN SRAM scrambling.
+ */
+static volatile uint32_t
+    sram_ctrl_main_scramble_buffer[SRAM_CTRL_TESTUTILS_DATA_NUM_WORDS];
+
+/**
+ * Test pattern to be written and read from SRAM.
+ */
+static const sram_ctrl_testutils_data_t kRamTestPattern1 = {
+    .words =
+        {
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+        },
+};
+
+/**
+ * Test pattern to be written and read from SRAM.
+ */
+static const sram_ctrl_testutils_data_t kRamTestPattern2 = {
+    .words =
+        {
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+        },
+};
+
+/**
+ * Performs scrambling, saves the test relevant data and resets the system.
+ *
+ * MAIN SRAM will be overwritten with the pseudo-random data (including .bss
+ * and .data segments), which means that the code must be written in assembly.
+ *
+ * This function saves contents of the `sram_ctrl_main_scramble_buffer` in the
+ * RETENTION SRAM, which which is kept intact across the system reboot.
+ */
+static noreturn void main_sram_scramble(void) {
+  asm volatile(
+      // Request the new scrambling key for MAIN SRAM.
+      "li %[tmp], 0x1;                                                     \n"
+      "sw %[tmp], %[kSramCtrlOffset](%[kSramCtrlRegsBase]);                \n"
+
+      // Busy loop - waiting for scrambling to finish.
+      ".L_scrambling_busy_loop:;                                           \n"
+      "  lw %[tmp], %[kSramCtrlStatusOffset](%[kSramCtrlRegsBase]);        \n"
+      "  andi %[tmp], %[tmp], %[kSramCtrlKeyScrDone];                      \n"
+      "  beqz %[tmp], .L_scrambling_busy_loop;                             \n"
+
+      // Copy the buffer contents to the retention SRAM.
+      ".L_buffer_copy_loop:;                                               \n"
+      "  lw %[tmp], 0(%[kCopyFromAddr]);                                   \n"
+      "  sw %[tmp], 0(%[kCopyToAddr]);                                     \n"
+      "  addi %[kCopyFromAddr], %[kCopyFromAddr], 4;                       \n"
+      "  addi %[kCopyToAddr], %[kCopyToAddr], 4;                           \n"
+      "  blt %[kCopyToAddr], %[kCopyToEndAddr], .L_buffer_copy_loop;       \n"
+
+      // Trigger the software system reset via the Reset Manager.
+      "li %[tmp], %[kMultiBitTrue];                                        \n"
+      "sw %[tmp], %[kRstMgrResetReq](%[kRstMgrRegsBase]);                  \n"
+
+      // Satisfy the `noreturn` promise to the compiler.
+      ".L_infinite_loop:;                                                  \n"
+      "  wfi;                                                              \n"
+      "  j .L_infinite_loop;"
+      :
+      // Allocating a register for general use.
```

Comment:
Maybe place this after the constraint for `tmp` so that clang-format does not include the immediate constraint after it in the same line.